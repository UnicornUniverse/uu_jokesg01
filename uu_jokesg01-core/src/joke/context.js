//@@viewOn:imports
import { Utils } from "uu5g05";
//@@viewOff:imports

export const [Context, useJoke] = Utils.Context.create(null);
export default Context;
