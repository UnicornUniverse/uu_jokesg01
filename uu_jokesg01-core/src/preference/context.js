//@@viewOn:imports
import { Utils } from "uu5g05";
//@@viewOff:imports

export const [Context, usePreference] = Utils.Context.create(null);
export default Context;
