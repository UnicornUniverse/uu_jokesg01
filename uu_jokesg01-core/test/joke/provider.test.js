import { Client } from "uu_appg01";
import { omitConsoleLogs, wait } from "uu5g05-test";
import { render } from "../tools";
import { useJoke } from "../../src/joke/context.js";
import Provider from "../../src/joke/provider.js";

const categoryList = [
  { id: "1", name: "Category 1" },
  { id: "2", name: "Category 2" },
];

const updatedCategoryList = [
  { id: "7", name: "Category 7" },
  { id: "8", name: "Category 8" },
];

const joke = {
  id: "joke-1",
  name: "Test joke",
  visibility: false,
  text: "Best joke ever",
  ratingCount: 5,
  averageRating: 2.5,
  categoryIdList: ["1", "2"],
  sys: {
    cts: "2020-01-01T00:00:00.000Z",
  },
};

const jokeWithImage = { ...joke, id: "joke-with-image-1", image: "image-1" };
const anotherJoke = { ...joke, id: "another-joke-1" };

function getDefaultProps() {
  return { baseUri: "http://localhost", oid: "joke-1", children: jest.fn() };
}

const ContextReader = ({ children }) => {
  const value = useJoke();
  return children(value);
};

async function setup(props = getDefaultProps()) {
  global.URL.createObjectURL = jest.fn(() => "http://localhost/123");
  global.URL.revokeObjectURL = jest.fn();

  const view = render(<Provider {...props} />);
  await wait();
  return { props, view };
}

describe("UuJokesCore.Joke.Provider", () => {
  it(`reads data through children as a function`, async () => {
    Client.get.mockReturnValueOnce({ data: joke }); // joke/get
    Client.get.mockReturnValueOnce({ data: { itemList: categoryList } }); // category/list
    const { props } = await setup();
    expect(props.children.mock.lastCall[0]).toMatchSnapshot();
  });

  it(`reads data throw context`, async () => {
    Client.get.mockReturnValueOnce({ data: joke }); // joke/get
    Client.get.mockReturnValueOnce({ data: { itemList: categoryList } }); // category/list
    const props = getDefaultProps();
    const consumer = jest.fn(() => null);
    props.children = <ContextReader>{consumer}</ContextReader>;
    await setup(props);

    expect(consumer.mock.lastCall[0]).toMatchSnapshot();
  });

  it(`rerenders component with different oid`, async () => {
    Client.get.mockReturnValueOnce({ data: joke }); // joke/get
    Client.get.mockReturnValueOnce({ data: { itemList: categoryList } }); // category/list
    Client.get.mockReturnValueOnce({ data: anotherJoke }); // joke/get
    Client.get.mockReturnValueOnce({ data: { itemList: categoryList } }); // category/list

    const { view, props } = await setup();

    view.rerender(<Provider {...props} oid="another-joke-1" />);
    await wait();

    expect(props.children.mock.lastCall[0]).toMatchSnapshot();
  });

  it(`reload with error`, async () => {
    Client.get.mockReturnValueOnce({ data: joke }); // joke/get
    Client.get.mockReturnValueOnce({ data: { itemList: categoryList } }); // category/list
    const { view, props } = await setup();

    Provider.logger.error = jest.fn();
    Client.get.mockImplementationOnce(() => {
      throw new Error("Test reload with error");
    }); // joke/get

    view.rerender(<Provider {...props} oid={"789"} />);
    await wait();

    expect(Provider.logger.error).toBeCalledTimes(1);
    expect(Provider.logger.error.mock.lastCall).toMatchSnapshot("logger");
    expect(props.children.mock.lastCall[0]).toMatchSnapshot("children");
  });

  it(`loads joke with image`, async () => {
    Client.get.mockReturnValueOnce({ data: joke }); // joke/get
    Client.get.mockReturnValueOnce({ data: { itemList: categoryList } }); // category/list
    Client.get.mockReturnValueOnce(new Blob()); // uu-app-binarystore/getBinaryData

    const props = { ...getDefaultProps(), oid: jokeWithImage.id };
    await setup(props);

    expect(props.children.mock.lastCall[0]).toMatchSnapshot();
  });

  it(`renders component without required oid property`, async () => {
    const props = { ...getDefaultProps(), oid: undefined };
    omitConsoleLogs("The prop `oid` is marked as required");
    omitConsoleLogs("NoOidError: The required property oid is not defined!");
    await setup(props);

    expect(props.children.mock.lastCall[0]).toMatchSnapshot("children");
  });

  it(`updates joke`, async () => {
    Client.get.mockReturnValueOnce({ data: jokeWithImage }); // joke/get
    Client.get.mockReturnValueOnce({ data: { itemList: categoryList } }); // category/list
    Client.get.mockReturnValueOnce(new Blob()); // uu-app-binarystore/getBinaryData
    Client.post.mockImplementationOnce((uri, dtoIn) => ({ data: { ...jokeWithImage, ...dtoIn } })); // joke/update
    Client.get.mockReturnValueOnce({ data: { itemList: updatedCategoryList } }); // category/list

    const props = getDefaultProps();
    props.children.mockImplementation(
      ({ jokeDataObject }) =>
        Client.post.mock.calls.length === 0 &&
        jokeDataObject.state === "ready" &&
        jokeDataObject.handlerMap.update({
          name: "New name",
          text: "New text",
          categoryIdList: ["7", "8"],
          image: "456",
        })
    );

    await setup(props);

    expect(props.children.mock.lastCall[0]).toMatchSnapshot();
  });

  it(`addRating`, async () => {
    Client.get.mockReturnValueOnce({ data: jokeWithImage }); // joke/get
    Client.get.mockReturnValueOnce({ data: { itemList: categoryList } }); // category/list
    Client.get.mockReturnValueOnce(new Blob()); // uu-app-binarystore/getBinaryData
    Client.post.mockReturnValueOnce({ data: { ...jokeWithImage, ratingCount: 6, averageRating: 3 } }); // joke/update

    const props = getDefaultProps();
    props.children.mockImplementation(
      ({ jokeDataObject }) =>
        Client.post.mock.calls.length === 0 &&
        jokeDataObject.state === "ready" &&
        jokeDataObject.handlerMap.addRating(5)
    );

    await setup(props);

    expect(props.children.mock.lastCall[0]).toMatchSnapshot();
  });

  it(`updates visibility`, async () => {
    Client.get.mockReturnValueOnce({ data: jokeWithImage }); // joke/get
    Client.get.mockReturnValueOnce({ data: { itemList: categoryList } }); // category/list
    Client.get.mockReturnValueOnce(new Blob()); // uu-app-binarystore/getBinaryData
    Client.post.mockImplementationOnce((uri, dtoIn) => ({
      data: { ...jokeWithImage, visibility: dtoIn.visibility },
    })); // joke/update

    const props = getDefaultProps();
    props.children.mockImplementation(
      ({ jokeDataObject }) =>
        Client.post.mock.calls.length === 0 &&
        jokeDataObject.state === "ready" &&
        jokeDataObject.handlerMap.updateVisibility(true)
    );

    await setup(props);

    expect(props.children.mock.lastCall[0]).toMatchSnapshot();
  });
});
