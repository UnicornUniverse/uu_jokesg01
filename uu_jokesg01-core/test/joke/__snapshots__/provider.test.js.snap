// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UuJokesCore.Joke.Provider addRating 1`] = `
Object {
  "jokeDataObject": Object {
    "data": Object {
      "averageRating": 3,
      "categoryIdList": Array [
        "1",
        "2",
      ],
      "id": "joke-with-image-1",
      "image": "image-1",
      "imageFile": undefined,
      "imageUrl": "http://localhost/123",
      "name": "Test joke",
      "ratingCount": 6,
      "sys": Object {
        "cts": "2020-01-01T00:00:00.000Z",
      },
      "text": "Best joke ever",
      "visibility": false,
    },
    "errorData": null,
    "handlerMap": Object {
      "addRating": [Function],
      "load": [Function],
      "setData": [Function],
      "update": [Function],
      "updateVisibility": [Function],
    },
    "pendingData": null,
    "state": "ready",
  },
}
`;

exports[`UuJokesCore.Joke.Provider loads joke with image 1`] = `
Object {
  "jokeDataObject": Object {
    "data": Object {
      "averageRating": 2.5,
      "categoryIdList": Array [
        "1",
        "2",
      ],
      "id": "joke-with-image-1",
      "image": "image-1",
      "imageFile": undefined,
      "imageUrl": "http://localhost/123",
      "name": "Test joke",
      "ratingCount": 5,
      "sys": Object {
        "cts": "2020-01-01T00:00:00.000Z",
      },
      "text": "Best joke ever",
      "visibility": false,
    },
    "errorData": null,
    "handlerMap": Object {
      "addRating": [Function],
      "load": [Function],
      "setData": [Function],
      "update": [Function],
      "updateVisibility": [Function],
    },
    "pendingData": null,
    "state": "ready",
  },
}
`;

exports[`UuJokesCore.Joke.Provider reads data through children as a function 1`] = `
Object {
  "jokeDataObject": Object {
    "data": Object {
      "averageRating": 2.5,
      "categoryIdList": Array [
        "1",
        "2",
      ],
      "id": "joke-1",
      "name": "Test joke",
      "ratingCount": 5,
      "sys": Object {
        "cts": "2020-01-01T00:00:00.000Z",
      },
      "text": "Best joke ever",
      "visibility": false,
    },
    "errorData": null,
    "handlerMap": Object {
      "addRating": [Function],
      "load": [Function],
      "setData": [Function],
      "update": [Function],
      "updateVisibility": [Function],
    },
    "pendingData": null,
    "state": "ready",
  },
}
`;

exports[`UuJokesCore.Joke.Provider reads data throw context 1`] = `
Object {
  "jokeDataObject": Object {
    "data": Object {
      "averageRating": 2.5,
      "categoryIdList": Array [
        "1",
        "2",
      ],
      "id": "joke-1",
      "name": "Test joke",
      "ratingCount": 5,
      "sys": Object {
        "cts": "2020-01-01T00:00:00.000Z",
      },
      "text": "Best joke ever",
      "visibility": false,
    },
    "errorData": null,
    "handlerMap": Object {
      "addRating": [Function],
      "load": [Function],
      "setData": [Function],
      "update": [Function],
      "updateVisibility": [Function],
    },
    "pendingData": null,
    "state": "ready",
  },
}
`;

exports[`UuJokesCore.Joke.Provider reload with error: children 1`] = `
Object {
  "jokeDataObject": Object {
    "data": Object {
      "averageRating": 2.5,
      "categoryIdList": Array [
        "1",
        "2",
      ],
      "id": "joke-1",
      "name": "Test joke",
      "ratingCount": 5,
      "sys": Object {
        "cts": "2020-01-01T00:00:00.000Z",
      },
      "text": "Best joke ever",
      "visibility": false,
    },
    "errorData": Object {
      "data": undefined,
      "dtoIn": undefined,
      "error": [Error: Test reload with error],
      "operation": "load",
    },
    "handlerMap": Object {
      "addRating": [Function],
      "load": [Function],
      "setData": [Function],
      "update": [Function],
      "updateVisibility": [Function],
    },
    "pendingData": null,
    "state": "error",
  },
}
`;

exports[`UuJokesCore.Joke.Provider reload with error: logger 1`] = `
Array [
  "Error while reloading data.",
  [Error: Test reload with error],
]
`;

exports[`UuJokesCore.Joke.Provider renders component without required oid property: children 1`] = `
Object {
  "jokeDataObject": Object {
    "data": null,
    "errorData": Object {
      "data": undefined,
      "dtoIn": undefined,
      "error": [NoOidError: The required property oid is not defined!],
      "operation": "load",
    },
    "handlerMap": Object {
      "addRating": [Function],
      "load": [Function],
      "setData": [Function],
      "update": [Function],
      "updateVisibility": [Function],
    },
    "pendingData": null,
    "state": "errorNoData",
  },
}
`;

exports[`UuJokesCore.Joke.Provider rerenders component with different oid 1`] = `
Object {
  "jokeDataObject": Object {
    "data": Object {
      "averageRating": 2.5,
      "categoryIdList": Array [
        "1",
        "2",
      ],
      "id": "another-joke-1",
      "name": "Test joke",
      "ratingCount": 5,
      "sys": Object {
        "cts": "2020-01-01T00:00:00.000Z",
      },
      "text": "Best joke ever",
      "visibility": false,
    },
    "errorData": null,
    "handlerMap": Object {
      "addRating": [Function],
      "load": [Function],
      "setData": [Function],
      "update": [Function],
      "updateVisibility": [Function],
    },
    "pendingData": null,
    "state": "ready",
  },
}
`;

exports[`UuJokesCore.Joke.Provider updates joke 1`] = `
Object {
  "jokeDataObject": Object {
    "data": Object {
      "averageRating": 2.5,
      "categoryIdList": Array [
        "7",
        "8",
      ],
      "id": "joke-with-image-1",
      "image": "456",
      "imageFile": "456",
      "imageUrl": "http://localhost/123",
      "name": "New name",
      "ratingCount": 5,
      "sys": Object {
        "cts": "2020-01-01T00:00:00.000Z",
      },
      "text": "New text",
      "visibility": false,
    },
    "errorData": null,
    "handlerMap": Object {
      "addRating": [Function],
      "load": [Function],
      "setData": [Function],
      "update": [Function],
      "updateVisibility": [Function],
    },
    "pendingData": null,
    "state": "ready",
  },
}
`;

exports[`UuJokesCore.Joke.Provider updates visibility 1`] = `
Object {
  "jokeDataObject": Object {
    "data": Object {
      "averageRating": 2.5,
      "categoryIdList": Array [
        "1",
        "2",
      ],
      "id": "joke-with-image-1",
      "image": "image-1",
      "imageFile": undefined,
      "imageUrl": "http://localhost/123",
      "name": "Test joke",
      "ratingCount": 5,
      "sys": Object {
        "cts": "2020-01-01T00:00:00.000Z",
      },
      "text": "Best joke ever",
      "visibility": true,
    },
    "errorData": null,
    "handlerMap": Object {
      "addRating": [Function],
      "load": [Function],
      "setData": [Function],
      "update": [Function],
      "updateVisibility": [Function],
    },
    "pendingData": null,
    "state": "ready",
  },
}
`;
